funsor/domains.py

```diff
@@ -175,6 +175,37 @@ def bint(size):
     return Bint[size]
 
 
+class ProductDomain(Domain):
+
+    _type_cache = WeakValueDictionary()
+
+    def __getitem__(cls, arg_domains):
+        try:
+            return ProductDomain._type_cache[arg_domains]
+        except KeyError:
+            assert isinstance(arg_domains, tuple)
+            assert all(isinstance(arg_domain, Domain) for arg_domain in arg_domains)
+            subcls = type("Product_", (Product,), {"__args__": arg_domains})
+            ProductDomain._type_cache[arg_domains] = subcls
+            return subcls
+
+    def __repr__(cls):
+        return "Product[{}]".format(", ".join(map(repr, cls.__args__)))
+
+    @property
+    def __origin__(cls):
+        return Product
+
+    @property
+    def shape(cls):
+        return (len(cls.__args__),)
+
+
+class Product(tuple, metaclass=ProductDomain):
+    """like typing.Tuple, but works with issubclass"""
+    __args__ = NotImplemented
```

Comment:
Does the `__args__` name have something to do with the `typing` module? I would expect Python-builtin-style naming to correspond to some Python concept, and to use `._args` otherwise.

Reply:
Yes, it's a convention I copied from `typing` whose purpose seems to be to signal that `__args__` are to be accessed through a `typing` introspection API (e.g. `typing_extensions.get_args(t)`) rather than directly, perhaps because the internals of `typing` change a lot between releases.  In a future PR we can add `get_origin()` and `get_args()` for `Domain`s.